File: src/lib/pitono/src/base_setup.py
from typing import List, Dict, Any, Callable, Optional
from .pitono_types import ITestResourceConfiguration, ITestArtifactory

class BaseSetup:
    def __init__(
        self,
        features: List[str],
        actions: List[Any],
        checks: List[Any],
        setup_cb: Callable,
        initial_values: Any
    ):
        self.features = features
        self.actions = actions
        self.checks = checks
        self.setup_cb = setup_cb
        self.initial_values = initial_values
        self.fails = 0
        self.failed = False
        self.error: Optional[Exception] = None
        self.store: Any = None
        self.key = ""
        self.status: Optional[bool] = None
        self.artifacts: List[str] = []
    
    def add_artifact(self, path: str) -> None:
        if not isinstance(path, str):
            raise TypeError(
                f"[ARTIFACT ERROR] Expected string, got {type(path)}: {path}"
            )
        normalized_path = path.replace('\\', '/')
        self.artifacts.append(normalized_path)
    
    def to_obj(self) -> Dict[str, Any]:
        action_objs = []
        for action in self.actions:
            if action is not None and hasattr(action, 'to_obj'):
                action_objs.append(action.to_obj())
        
        check_objs = []
        for check in self.checks:
            if check is not None and hasattr(check, 'to_obj'):
                check_objs.append(check.to_obj())
        
        error_info = None
        if self.error:
            error_info = f"{type(self.error).__name__}: {str(self.error)}"
        
        return {
            'key': self.key,
            'actions': action_objs,
            'checks': check_objs,
            'error': error_info,
            'failed': self.failed,
            'features': self.features,
            'artifacts': self.artifacts,
            'status': self.status
        }
    
    async def setup_that(
        self,
        subject: Any,
        test_resource_configuration: ITestResourceConfiguration,
        artifactory: ITestArtifactory,
        setup_cb: Callable,
        initial_values: Any
    ) -> Any:
        raise NotImplementedError("setup_that must be implemented by subclasses")
    
    async def after_each(
        self,
        store: Any,
        key: str,
        artifactory: ITestArtifactory
    ) -> Any:
        return store
    
    async def setup(
        self,
        subject: Any,
        key: str,
        test_resource_configuration: ITestResourceConfiguration,
        tester: Callable[[Any], bool],
        artifactory: ITestArtifactory,
        suite_ndx: int
    ) -> Any:
        self.key = key
        self.fails = 0
        
        def setup_artifactory(f_path: str, value: Any):
            return artifactory(f"setup-{key}/{f_path}", value)
        
        self.store = await self.setup_that(
            subject,
            test_resource_configuration,
            setup_artifactory,
            self.setup_cb,
            self.initial_values
        )
        self.status = True
        
        # Process actions
        for action_ndx, action_step in enumerate(self.actions):
            # Create artifactory for action context
            action_artifactory = self._create_artifactory_for_action(key, action_ndx, suite_ndx)
            self.store = await action_step.test(
                self.store,
                test_resource_configuration,
                action_artifactory
            )
        
        # Process checks
        for check_ndx, check_step in enumerate(self.checks):
            if suite_ndx is not None:
                filepath = f"suite-{suite_ndx}/setup-{key}/check-{check_ndx}"
            else:
                filepath = f"setup-{key}/check-{check_ndx}"
            
            # Create artifactory for check context
            check_artifactory = self._create_artifactory_for_check(key, check_ndx, suite_ndx)
            result = await check_step.test(
                self.store,
                test_resource_configuration,
                filepath,
                check_artifactory
            )
            if not tester(result):
                self.failed = True
                self.fails += 1
        
        await self.after_each(self.store, self.key, setup_artifactory)
        
        return self.store
    
    def _create_artifactory_for_action(self, key: str, action_index: int, suite_ndx: Optional[int] = None):
        # This should be implemented by subclasses that have access to parent
        # For now, return a simple artifactory
        return lambda f_path, value: None
    
    def _create_artifactory_for_check(self, key: str, check_index: int, suite_ndx: Optional[int] = None):
        # This should be implemented by subclasses that have access to parent
        # For now, return a simple artifactory
        return lambda f_path, value: None

File: src/lib/kafe/src/main/java/kafe/BaseSetup.java
package kafe;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.function.Function;

/**
 * BaseSetup is the unified base class for all setup phases.
 * It covers BDD's Given, AAA's Arrange, and TDT's Map.
 */
public abstract class BaseSetup<I, S, R> {
    public List<String> features;
    public List<BaseAction<S, R>> actions;
    public List<BaseCheck<R>> checks;
    public Exception error;
    public Object fail;
    public R store;
    public String recommendedFsPath;
    public Function<S, R> setupCB;
    public Object initialValues;
    public String key;
    public boolean failed;
    public List<String> artifacts;
    public int fails;
    public Boolean status;
    
    public BaseSetup(
        List<String> features,
        List<BaseAction<S, R>> actions,
        List<BaseCheck<R>> checks,
        Function<S, R> setupCB,
        Object initialValues
    ) {
        this.features = features != null ? features : new ArrayList<>();
        this.actions = actions != null ? actions : new ArrayList<>();
        this.checks = checks != null ? checks : new ArrayList<>();
        this.setupCB = setupCB;
        this.initialValues = initialValues;
        this.fails = 0;
        this.failed = false;
        this.error = null;
        this.store = null;
        this.key = "";
        this.status = null;
        this.artifacts = new ArrayList<>();
    }
    
    public void addArtifact(String path) {
        if (!(path instanceof String)) {
            throw new RuntimeException(
                "[ARTIFACT ERROR] Expected string, got " + path.getClass().getName() + ": " + path
            );
        }
        String normalizedPath = path.replace('\\', '/');
        artifacts.add(normalizedPath);
    }
    
    public Map<String, Object> toObj() {
        List<Map<String, Object>> actionObjs = new ArrayList<>();
        for (BaseAction<S, R> action : actions) {
            if (action != null) {
                actionObjs.add(action.toObj());
            }
        }
        
        List<Map<String, Object>> checkObjs = new ArrayList<>();
        for (BaseCheck<R> check : checks) {
            if (check != null) {
                checkObjs.add(check.toObj());
            }
        }
        
        Map<String, Object> obj = new HashMap<>();
        obj.put("key", key);
        obj.put("actions", actionObjs);
        obj.put("checks", checkObjs);
        obj.put("error", error != null ? error.getMessage() : null);
        obj.put("failed", failed);
        obj.put("features", features);
        obj.put("artifacts", artifacts);
        obj.put("status", status);
        return obj;
    }
    
    public abstract R setupThat(
        S subject,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory,
        Function<S, R> setupCB,
        Object initialValues
    );
    
    public R afterEach(
        R store,
        String key,
        Object artifactory
    ) {
        return store;
    }
    
    public R setup(
        S subject,
        String key,
        ITTestResourceConfiguration testResourceConfiguration,
        Function<Object, Boolean> tester,
        Object artifactoryObj,
        int suiteNdx
    ) throws Exception {
        this.key = key;
        this.fails = 0;
        
        // Ensure we have an artifactory - create a default one if none provided
        Object setupArtifactory = artifactoryObj;
        if (setupArtifactory == null) {
            // Create a simple default artifactory that does nothing
            setupArtifactory = new Object() {
                public void writeFileSync(String filename, String payload) {
                    System.out.println("[Default Artifactory] Would write to: " + filename);
                }
            };
        }
        
        try {
            this.store = setupThat(
                subject,
                testResourceConfiguration,
                setupArtifactory,
                this.setupCB,
                this.initialValues
            );
            this.status = true;
        } catch (Exception e) {
            this.status = false;
            this.failed = true;
            this.fails++;
            this.error = e;
            return this.store;
        }
        
        try {
            // Process actions with artifactory
            for (int actionNdx = 0; actionNdx < actions.size(); actionNdx++) {
                BaseAction<S, R> actionStep = actions.get(actionNdx);
                try {
                    // Create context-specific artifactory for this action
                    Object actionArtifactory = createActionArtifactory(key, actionNdx, suiteNdx, artifactoryObj);
                    this.store = actionStep.performAction(
                        this.store,
                        actionStep.actionCB,
                        testResourceConfiguration,
                        actionArtifactory
                    );
                } catch (Exception e) {
                    this.failed = true;
                    this.fails++;
                    this.error = e;
                }
            }
            
            // Process checks with artifactory
            for (int checkNdx = 0; checkNdx < checks.size(); checkNdx++) {
                BaseCheck<R> checkStep = checks.get(checkNdx);
                try {
                    String filepath = suiteNdx != -1 ? 
                        "suite-" + suiteNdx + "/setup-" + key + "/check-" + checkNdx :
                        "setup-" + key + "/check-" + checkNdx;
                    
                    // Create context-specific artifactory for this check
                    Object checkArtifactory = createCheckArtifactory(key, checkNdx, suiteNdx, artifactoryObj);
                    Object result = checkStep.verifyCheck(
                        this.store,
                        checkStep.checkCB,
                        testResourceConfiguration,
                        checkArtifactory
                    );
                    if (!tester.apply(result)) {
                        this.failed = true;
                        this.fails++;
                    }
                } catch (Exception e) {
                    this.failed = true;
                    this.fails++;
                    this.error = e;
                }
            }
        } catch (Exception e) {
            this.error = e;
            this.failed = true;
            this.fails++;
        } finally {
            try {
                this.store = afterEach(this.store, this.key, setupArtifactory);
            } catch (Exception e) {
                this.failed = true;
                this.fails++;
                this.error = e;
            }
        }
        
        return this.store;
    }
    
    private Object createActionArtifactory(String key, int actionIndex, int suiteNdx, Object parentArtifactory) {
        // In a real implementation, this would create a context-specific artifactory
        // For now, return the parent artifactory
        return parentArtifactory;
    }
    
    private Object createCheckArtifactory(String key, int checkIndex, int suiteNdx, Object parentArtifactory) {
        // In a real implementation, this would create a context-specific artifactory
        // For now, return the parent artifactory
        return parentArtifactory;
    }
}
